compiler/src/dotty/tools/dotc/transform/DropBreaks.scala

```diff
@@ -0,0 +1,213 @@
+   *  @param  goto     the return-label to use for a labeled return.
+   *  @param  enclMeth the enclosing method
+   */
+  class LabelUsage(val goto: TermSymbol, val enclMeth: Symbol):
+    /** The number of references to associated label that come from labeled returns */
+    var returnRefs: Int = 0
+    /** The number of other references to assocated label */
+    var otherRefs: Int = 0
+
+  private val LabelUsages = new Property.Key[Map[Symbol, LabelUsage]]
+  private val LabelsShadowedByTry = new Property.Key[Set[Symbol]]
+
+/** Rewrites local Break throws to labeled returns.
+ *  Drops `try` statements on breaks if no other uses of its label remain.
+ *  A Break throw with a `Label` created by some enclosing boundary is replaced
+ *  with a labeled return if
+ *
+ *    - the throw and the boundary are in the same method, and
+ *    - there is no try expression inside the boundary that encloses the throw.
+ */
+class DropBreaks extends MiniPhase:
+  import DropBreaks.*
+
+  import tpd._
+
+  override def phaseName: String = DropBreaks.name
+
+  override def description: String = DropBreaks.description
+
+  override def runsAfterGroupsOf: Set[String] = Set(ElimByName.name)
+    // we want by-name parameters to be converted to closures
+
+  private object LabelTry:
+
+    object GuardedThrow:
+
+      /** `(ex, local)` provided `expr` matches
+       *
+       *      if ex.label.eq(local) then ex.value else throw ex
+       */
+      def unapply(expr: Tree)(using Context): Option[(Symbol, Symbol)] = stripTyped(expr) match
+        case If(
+          Apply(Select(Select(ex: Ident, label), eq), (lbl @ Ident(local)) :: Nil),
+          Select(ex2: Ident, value),
+          Apply(throww, (ex3: Ident) :: Nil))
+        if label == nme.label && eq == nme.eq && local == nme.local && value == nme.value
+            && throww.symbol == defn.throwMethod
+            && ex.symbol == ex2.symbol && ex.symbol == ex3.symbol =>
+          Some((ex.symbol, lbl.symbol))
+        case _ =>
+          None
+    end GuardedThrow
+
+    /** `(local, body)` provided `tree` matches
+     *
+     *      try body
+     *      catch case ex: Break =>
+     *        if ex.label.eq(local) then ex.value else throw ex
+     */
+    def unapply(tree: Tree)(using Context): Option[(Symbol, Tree)] = stripTyped(tree) match
+      case Try(body, CaseDef(pat @ Bind(_, Typed(_, tpt)), EmptyTree, GuardedThrow(exc, local)) :: Nil, EmptyTree)
+      if tpt.tpe.isRef(defn.BreakClass) && exc == pat.symbol =>
+        Some((local, body))
+      case _ =>
+        None
+  end LabelTry
+
+  private object BreakBoundary:
+
+    /** `(local, body)` provided `tree` matches
+     *
+     *      { val local: Label[...] = ...; <LabelTry(local, body)> }
+     */
+    def unapply(tree: Tree)(using Context): Option[(Symbol, Tree)] = stripTyped(tree) match
+      case Block((vd @ ValDef(nme.local, _, _)) :: Nil, LabelTry(caughtAndRhs))
```

Comment:
Can't the `ValDef` be renamed under inlining? It might not always be called `nme.local`.

Reply:
Currently I don't see how it would be renamed. Testing for the name means we narrow down the check must faster, so we can execute `LabelTry` only for very likely positives.